File: app/ios/Kordi/Features/Conversation/SubsessionConversation.swift
import Foundation

extension CloudAgentSubsession {
    var conversation: ConversationSummary {
        ConversationSummary(id: "subsession:\(sessionId)", kind: .agent,
            peerAccountId: ownerAccountId, agentId: agentId, ownerDisplayName: ownerDisplayName,
            displayName: title, lastMessage: messages.last?.text ?? "", lastActivityAt: .distantPast,
            unreadCount: 0, avatarSource: agentAvatarUrl, agentActivity: state == .running ? .replying : .ready,
            sessionId: sessionId, agentDisplayName: agentDisplayName,
            groupParticipants: (participants ?? []).map {
                CloudGroupParticipant(accountId: $0.accountId, displayName: $0.displayName,
                    avatarUrl: $0.avatarUrl, role: nil)
            }, subsessionId: sessionId)
    }

    func mentionTargets(accountId: String) -> [ComposerMentionTarget] {
        [ComposerMentionTarget(id: agentId, displayName: agentDisplayName, kind: .agent,
            accountId: ownerAccountId, agentId: agentId, ownerName: ownerDisplayName, avatarSource: agentAvatarUrl)]
        + (participants ?? []).filter { $0.accountId != accountId }.map {
            ComposerMentionTarget(id: $0.accountId, displayName: $0.displayName, kind: .person,
                accountId: $0.accountId, agentId: nil, ownerName: nil, avatarSource: $0.avatarUrl)
        }
    }

    func chatMessages(accountId: String) -> [ChatMessage] {
        var result: [ChatMessage] = []
        var queuePosition = 0
        for message in messages {
            let agent = message.role == "assistant"
            if agent && ["queued", "pending", "leased"].contains(message.requestState ?? "") { continue }
            let author: MessageAuthor = agent ? .agent : message.senderAccountId == accountId ? .me : .person
            let phase: AgentExecutionSnapshot.Phase? = switch message.requestState {
                case "running" where agent: .usingTool
                case "completed" where agent: .complete
                case "failed" where agent: .failed
                case "cancelled" where agent: .cancelled
                default: nil
            }
            let execution = phase.map { phase in
                AgentExecutionSnapshot(phase: phase, summary: "", steps: [], tools: message.activity?.tools,
                    startedAtMs: nil, updatedAtMs: Double(message.timestampMs), completed: phase != .usingTool)
            }
            var row = ChatMessage(id: message.id, conversationId: conversation.id, author: author,
                authorName: agent ? agentDisplayName : author == .me ? "You" : message.senderDisplayName ?? "Task",
                senderOwnerName: agent ? ownerAccountId == accountId ? "You" : ownerDisplayName : nil,
                text: message.text, createdAt: Date(timeIntervalSince1970: Double(message.timestampMs) / 1000),
                deliveryState: .delivered, errorMessage: nil, requestMessageId: message.requestId,
                mentions: message.mentions ?? [], agentExecution: execution)
            if !agent && message.requestState == "queued" {
                queuePosition += 1
                row.agentQueuePosition = queuePosition
            }
            result.append(row)
        }
        if state == .running && hasFollowupExecution != true {
            let progress = AgentExecutionSnapshot(phase: .usingTool, summary: "", steps: [], tools: activity?.tools,
                startedAtMs: nil, updatedAtMs: 0, completed: false)
            if let index = result.lastIndex(where: { $0.author == .agent && $0.requestMessageId == nil }) {
                result[index].agentExecution = progress
            } else {
                result.insert(ChatMessage(id: "runtime:\(sessionId)", conversationId: conversation.id,
                    author: .agent, authorName: agentDisplayName,
                    senderOwnerName: ownerAccountId == accountId ? "You" : ownerDisplayName,
                    text: "", createdAt: .distantPast, deliveryState: .delivered, errorMessage: nil,
                    requestMessageId: nil, agentExecution: progress), at: 0)
            }
        }
        return result
    }
}

File: app/desktop/src/features/cloud/subsessionConversation.ts
import type { CloudAgentSubsession, AgentSubsessionMessage } from './agentSubsessionTypes';
import type { Conversation, Message, MessageMention, DesktopChatToolSnapshot } from '@/kordi-app/types';
import type { ComposerMentionOption } from '@/kordi-app/components/composer';
import { publicScopedAgentMentionHandle } from '@/lib/identityLabels';
import { mentionHandleForLabel } from '@/features/chat/messageActions/mentionHandles';

export function subsessionMentionOptions(record: CloudAgentSubsession, accountId: string): ComposerMentionOption[] {
  return [{
    value: publicScopedAgentMentionHandle(record.ownerDisplayName, record.agentDisplayName),
    label: record.agentDisplayName, detail: `Owner · ${record.ownerAccountId === accountId ? 'You' : record.ownerDisplayName}`,
    targetKind: 'agent', agentId: record.agentId, humanId: record.ownerAccountId,
    sourceHostId: 'cloud', nodeId: record.agentId, runtime: 'cloud', ownerName: record.ownerDisplayName,
    avatarImageUrl: record.agentAvatarUrl,
  }, ...(record.participants ?? []).filter(person => person.accountId !== accountId).map(person => ({
    value: mentionHandleForLabel(person.displayName), label: person.displayName, targetKind: 'person' as const,
    humanId: person.accountId, sourceHostId: 'cloud', nodeId: person.accountId, runtime: 'cloud',
    avatarImageUrl: person.avatarUrl, avatarSeed: person.avatarSeed,
  }))];
}

export function subsessionMentions(text: string, options: ComposerMentionOption[]): MessageMention[] {
  return Array.from(text.matchAll(/(?<![\p{L}\p{N}@])@([\p{L}\p{N}]+)/gu)).flatMap(match => {
    const candidates = options.filter(option => option.value === match[1]);
    // Ambiguous display handles must never select an identity by position.
    if (candidates.length !== 1) return [];
    const option = candidates[0];
    return [{ label: option.value, targetKind: option.targetKind, targetIdentityId: option.agentId ?? option.humanId,
      agentId: option.agentId, humanId: option.humanId, displayText: match[0], displayLabel: option.label,
      startUtf16: match.index, lengthUtf16: match[0].length }];
  });
}

export function subsessionTranscript(record: CloudAgentSubsession, accountId: string): Message[] {
  const rows = record.messages.filter(row => !(row.role === 'assistant' && (
    ['queued', 'pending', 'leased'].includes(row.requestState ?? '')
      || record.live === false && row.requestState === 'running' && !row.text.trim()
  )));
  const result = rows.map((row): Message => {
    const assistant = row.role === 'assistant';
    const agent = assistant || row.senderAgentId === record.agentId;
    const own = row.senderAccountId === accountId;
    const participant = record.participants?.find(person => person.accountId === row.senderAccountId);
    return {
      id: row.id, role: agent ? record.ownerAccountId === accountId ? 'owned-agent' : 'external-agent' : own ? 'user' : 'person',
      sender: agent ? record.agentDisplayName : own ? 'You' : row.senderDisplayName ?? 'Task',
      senderOwnerName: agent ? record.ownerAccountId === accountId ? 'You' : record.ownerDisplayName : undefined,
      senderIdentityId: agent ? record.agentId : row.senderAccountId,
      senderAvatarSeed: agent ? record.agentId : participant?.avatarSeed ?? row.senderAccountId,
      senderProfileImageUrl: agent ? record.agentAvatarUrl : participant?.avatarUrl,
      senderType: agent ? 'agent' : 'human', isOwnMessage: !agent && own, showSenderMeta: true,
      text: row.text, timestampMs: row.timestampMs,
      statusChips: !agent && row.requestState === 'queued' ? ['queued'] : undefined,
      time: new Date(row.timestampMs).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }),
      mentions: row.mentions, turn: assistant && row.requestState && !(record.live === false && row.requestState === 'running') ? execution(row, row.activity?.tools) : undefined,
      replyToMessageId: assistant ? row.requestId : undefined,
    };
  });
  if (record.status === 'running' && record.live !== false && !record.hasFollowupExecution) {
    const reversed = [...rows].reverse();
    const lastAnswer = reversed.find(row => row.role === 'assistant' && !row.requestId);
    const last = result.find(row => row.id === lastAnswer?.id);
    const taskBrief = reversed.find(row => row.senderAgentId === record.agentId);
    const progress = execution({ id: `runtime:${record.sessionId}`, role: 'assistant', text: '', timestampMs: 0, requestState: 'running' }, record.activity?.tools);
    if (last) last.turn = { ...progress, assistantText: last.text };
    else result.splice(result.findIndex(row => row.id === taskBrief?.id) + 1, 0,
      { id: progress.id, role: record.ownerAccountId === accountId ? 'owned-agent' : 'external-agent', senderType: 'agent', sender: record.agentDisplayName,
      senderOwnerName: record.ownerAccountId === accountId ? 'You' : record.ownerDisplayName,
      senderIdentityId: record.agentId, senderProfileImageUrl: record.agentAvatarUrl,
      senderAvatarSeed: record.agentId, text: '', time: '', turn: progress });
  }
  return result;
}

export function subsessionConversation(id: string, record: CloudAgentSubsession | null, accountId: string): Conversation {
  return {
    id, canonicalSessionId: id, agentSubsessionId: id,
    name: record?.title ?? 'Agent session',
    type: record?.ownerAccountId === accountId ? 'owned-agent' : 'external-agent',
    subtitle: record ? `${record.agentDisplayName} · Owner · ${record.ownerAccountId === accountId ? 'You' : record.ownerDisplayName}` : 'Loading conversation…',
    unread: 0, collaborationSources: ['Cloud'], trust: 'Shared', directness: 'Agent session',
    participants: record ? (record.participants ?? []).map(person => person.displayName) : [],
    canonicalParticipants: record ? (record.participants ?? []).map(person => ({
      id: person.accountId, humanId: person.accountId, sourceIdentityId: person.accountId,
      name: person.displayName, kind: 'human' as const, source: 'cloud',
      role: person.accountId === accountId ? 'self' : 'participant',
      avatarKey: person.avatarSeed, profileImageUrl: person.avatarUrl,
    })) : [],
    avatarSeed: record?.agentId,
    profileImageUrl: record?.agentAvatarUrl,
    messages: record ? subsessionTranscript(record, accountId) : [],
  };
}

function execution(row: AgentSubsessionMessage, tools: DesktopChatToolSnapshot[] = []): NonNullable<Message['turn']> {
  const completed = ['completed', 'failed', 'cancelled'].includes(row.requestState ?? '');
  return { id: row.id, sessionId: '', prompt: '', status: row.requestState ?? 'running', message: '',
    assistantText: row.text, thinkingText: '', tools, completed, succeeded: row.requestState === 'completed', replyToMessageId: row.requestId };
}
